src/MessageHandler/Consumer.php

```diff
@@ -141,4 +108,84 @@ public function ack(Message $message): int
 
         return $r;
     }
+
+    /**
+     * the working list has to be unique per consumer worker
+     */
+    protected function getWorkingList(): string
+    {
+        return $this->queue->getWorkingListPrefixName().'-'.$this->uniqueId;
+    }
+
+    /**
+     * @param array $lists
+     * @param int   $nbMessage
+     *
+     * @return \SplStack
+     */
+    protected function getMessageFromQueueLists(array $lists, int $nbMessage): \SplStack
+    {
+        $messages = new \SplStack();
+
+        foreach ($lists as $list) {
+            while ($messages->count() < $nbMessage && $message = $this->redisClient->rpop($list)) {
+                $messages->push(self::unserializeMessage($message));
+            }
+        }
+
+        return $messages;
+    }
+
+    /**
+     * messages have to be acked manually via ->ack(message)
+     *
+     * - Get message from working list
+     *
+     * - grab something on another working list. Retrieve from working list on another instance of Consumer
+     * this is done only if the message is old enough to be considered as lost
+     *
+     * - get messages from queue list and add them in working queue
+     *
+     * @param array $lists
+     * @param int   $nbMessage
+     *
+     * @return \SplStack
+     */
+    protected function getMessagesFromWorkingLists(array $lists, int $nbMessage): \SplStack
+    {
+        $messages = new \SplStack();
+
+        // anything in the working list ? if so grab element stay on working list (not ack)
+        foreach ($this->redisClient->lrange($this->getWorkingList(), 0, $nbMessage - 1) as $message) {
+            $messages->push(self::unserializeMessage($message));
+        }
+
+        // wohaaa - nothing in the working list !
+        // grab something on another working list. Retrieve from working list on another instance of Consumer
+        // this is done only if the message is old enough to be considered as lost
+        if ($this->timeOldMessage && $messages->count() < $nbMessage) {
+            foreach ($this->queue->getWorkingLists($this->redisClient) as $list) {
+                foreach ($this->redisClient->lrange($list, 0, ($nbMessage - $messages->count() - 1)) as $message) {
+                    $message = self::unserializeMessage($message);
+                    if ((time() - $message->getCreatedAt()->format('U')) > $this->timeOldMessage) {
+                        $messages->push($message);
```

Comment:
When a message is ack, it's ack on all working queue, but i need to update message date.